apistar/http.py

```diff
@@ -1,263 +1,30 @@
-            value = validate(cls.schema, value, key=arg_name)
-        return value
-
-
-class Request(object):
-    __slots__ = ('method', 'url', 'headers')
-
-    def __init__(self, method: str, url: str, headers: HeadersType=None) -> None:
-        self.method = method
-        self.url = url
-        self.headers = Headers(headers)
-
-    @classmethod
-    def build(cls,
-              method: Method,
-              url: URL,
-              headers: Headers):
-        return cls(method=method, url=url, headers=headers)
-
-
-class Response(object):
-    __slots__ = ('data', 'content', 'status', 'headers')
-
+import collections
+import typing
+
+Method = typing.NewType('Method', str)
+URL = typing.NewType('URL', str)
+Scheme = typing.NewType('Scheme', str)
+Host = typing.NewType('Host', str)
+Port = typing.NewType('Port', int)
+Path = typing.NewType('Path', str)
+QueryString = typing.NewType('QueryString', str)
+QueryParams = typing.NewType('QueryParams', typing.Dict[str, str])
+QueryParam = typing.NewType('QueryParam', str)
+Headers = typing.NewType('Headers', typing.Dict[str, str])
+Header = typing.NewType('Header', str)
+Body = typing.NewType('Body', bytes)
+
+RequestData = typing.TypeVar('RequestData')
+
+
+class Response(collections.abc.Iterable):
     def __init__(self,
-                 data: Any,
-                 status: int=None,
-                 headers: HeadersType=None) -> None:
-        if headers is None:
-            headers_dict = {}  # type: Union[Dict[str, str], Headers]
-            headers_list = []  # type: List[Tuple[str, str]]
-        elif isinstance(headers, dict):
-            headers_dict = headers
-            headers_list = list(headers.items())
-        elif isinstance(headers, list):
-            headers_dict = dict(headers)
-            headers_list = headers
-        else:
-            headers_dict = headers
-            headers_list = headers.to_list()
-
-        if isinstance(data, str):
-            content = data.encode('utf-8')
-            if content:
-                content_type = 'text/html; charset=utf-8'
-            else:
-                content_type = None
-        elif isinstance(data, bytes):
-            content = data
-            if content:
-                content_type = 'text/html; charset=utf-8'
-            else:
-                content_type = None
-        else:
-            content = json.dumps(data).encode('utf-8')
-            content_type = 'application/json'
-
-        if 'Content-Length' not in headers_dict:
-            headers_list += [('Content-Length', str(len(content)))]
-        if content_type and 'Content-Type' not in headers_dict:
-            headers_list += [('Content-Type', content_type)]
-
-        if status is None:
-            status = 200 if content_type else 204
-
-        self.data = data
+                 content: typing.Any,
+                 status: int=200,
+                 headers: typing.Dict[str, str]=None) -> None:
```

Comment:
We'll likely want this annotation to actually be a `Union` of `dict/list of tuples/Headers`.
(ie. let the user use any of those, and then coerce them to Headers.
And yes, don't consider that annotation to be complete just yet, will get updated before 0.2 lands.